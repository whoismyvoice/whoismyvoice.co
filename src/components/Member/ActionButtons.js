import React from 'react';
import SenateStore from '../../stores/SenateStore';

// Components
import BaseComponent from '../BaseComponent';
import ContactButtonSmall from './../Buttons/ContactButtonSmall';

// Styles
import style from './../../styles/ActionButtons.scss';

class SupportActions extends BaseComponent {
  constructor(props) {
    super(props);
    this.state = SenateStore.getMember();
  }

  render() {
    let gender,
      email,
      twitter,
      tel,
      twitterLink;

    const {representative} = this.props;

    // Define each value used for every member
    if (representative) {
      gender = representative.gender === 'M' ? 'Him' : 'Her';
      email = representative.oc_email;
      twitter = representative.twitter_id;
      tel = representative.phone;
      twitterLink = `https://twitter.com/intent/tweet?hashtags=WhoIsMyVoice&text=@${twitter}`;
    }

    return  <div className="actionButtons">
      <ContactButtonSmall
        link={ 'mailto:' + email }
        icon="email"
        text={`Email ${gender}`}
        add_style="smaller"
      />
      <ContactButtonSmall
        link={`tel:${tel}`}
        icon="phone"
        text={`Call ${gender}`}
      />
      <ContactButtonSmall
        link={twitterLink}
        icon="twitter"
        text={`Tweet at ${gender}`}
      />
    </div>;
  }
}

SupportActions.propTypes = {
  representative: React.PropTypes.any
};

export default SupportActions;
