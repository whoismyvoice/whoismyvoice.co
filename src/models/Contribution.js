// @flow

import PropTypes from 'prop-types';

/**
 * Defines the type alias for a `Contribution` record.
 */
export type ContributionRecord = {
  amount: number,
  legislatorId: string,
  organization: string,
};

/**
 * Defines the `PropTypes` validation structure for a `Contribution` record.
 */
export const PropType = PropTypes.shape({
  amount: PropTypes.number.isRequired,
  legislatorId: PropTypes.string.isRequired,
  organization: PropTypes.string.isRequired,
});
