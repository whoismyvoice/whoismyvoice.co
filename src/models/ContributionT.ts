/**
 * Defines the type alias for a `Contribution` record.
 */
export type Record = {
  amount: number;
  legislatorId: string;
  organization: string;
};
