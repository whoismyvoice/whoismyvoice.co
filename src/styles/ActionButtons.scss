@import '_variables';
@import '_transitions';

.actionButtons {
  width: 100%;
  margin: 0 auto;
  display: inline-block;
  margin-top: -10px;

  @media only screen and (min-device-width: 768px) and (max-device-width: 1024px) and (orientation: landscape) and (-webkit-min-device-pixel-ratio: 1) {
    margin-top: 10px;
  }

  @media only screen and (min-device-width: 768px) and (max-device-width: 1024px) and (orientation: portrait) and (-webkit-min-device-pixel-ratio: 1) {
    margin-top: 20px;
  }

  @media screen and (min-width: 1600px) {
    margin-top: 15px;
  }

  @media screen and (max-width: $iphone6) {
    display: none;
  }
}

.mobile-contact-options {
  width: 70%;
  padding: 6px 5px 5px 5px;
  background-color: rgba(255, 255, 255, 0.1);
  border: 2px solid rgba(255, 255, 255, 0.4);
  color: $white;
  height: 25px;
  font-size: 14px;
  border-radius: 4px;
  font-family: 'Dosis';
  font-weight: 600;
  display: none;
  margin: 0 auto;
  position: relative;

  @media screen and (max-width: $iphone5-min) {
    font-size: 1.1em;
  }

  &:before {
    content: '';
    height: 40px;
    width: 40px;
    margin-top: -5px;
    margin-left: -2px;
    background-image: url('../img/icons/phone.png');
    background-position: center;
    background-size: 100%;
    display: block;
    float: left;
    position: relative;
    border-radius: 50%;

    @media screen and (max-width: $iphone5-min) {
      height: 30px;
      margin-top: -2px;
      width: 30px;
    }
  }

  @media screen and (max-width: $iphone6) {
    display: block;
  }
}

.mobile-contact-overlay {
  @include transition(0.3s);
  position: absolute;
  top: -30vh;
  z-index: 999;
  left: 0;
  display: flex;
  visibility: hidden;
  display: none;
  opacity: 0;
  width: 100vw;
  height: 100vh;
  background-color: rgba(255, 149, 73, 0.9);

  @media screen and (max-width: $iphone6) {
    display: block;
  }

  @media screen and (max-width: $iphone5-min) {
    top: -35vh;
  }

  .actionButtons {
    @media screen and (max-width: $iphone6) {
      display: block;
      margin-top: 40%;
      width: 100%;
    }

    @media screen and (max-width: $iphone5-min) {
      margin-top: 50%;
    }

    .contact__button_small {
      width: auto;
      padding: 1px 8px 0 5px;
      margin-bottom: 20px;
      display: block;
      margin: 0 auto;
    }

    .button__small__text {
      margin-top: 7px;
    }
  }

  &.show {
    @include transition(0.3s);
    justify-content: center;
    align-content: center;
    opacity: 1;
    visibility: visible;
    justify-content: center;

    .contact__button_small {
      margin: 0 auto;
      margin-bottom: 20px;
      display: block;
    }
  }
}

.mobile-contact-close-button {
  width: 35px;
  height: 35px;
  background-color: $white;
  margin: 0 auto;
  margin-top: 10%;
  border-radius: 50%;
  background-image: url('../img/icons/plus_icon.png');
  background-size: 50%;
  background-repeat: no-repeat;
  background-position: center center;
  transform: rotate(45deg);
  -moz-box-shadow: 3px 3px 5px rgba(68, 68, 68, 0.2);
  -webkit-box-shadow: 3px 3px 5px rgba(68, 68, 68, 0.2);
  box-shadow: 3px 3px 5px rgba(68, 68, 68, 0.2);
}
