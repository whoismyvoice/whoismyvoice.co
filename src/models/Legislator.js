// @flow

import PropTypes from 'prop-types';
import { PropType as ContributionType } from './Contribution';

// Types
import type { ContributionRecord } from './Contribution';

export type BioguideId = string;
export type FecId = string;
opaque type LegislatorChamber = 'senate' | 'house' | 'unknown';
export type LegislatorIdentifier = string;

type LegislatorBioRecord = {
  birthday: string,
  gender: string,
};

export type LegislatorChannelRecord = {
  id: string,
  type: string,
};

type LegislatorIdRecord = {
  bioguide: BioguideId,
  fec: Array<FecId>,
};

type LegislatorNameRecord = {
  first?: string,
  last?: string,
  official_full: string,
};

type LegislatorTermRecord = {
  type: string,
  start: string,
  end: string,
  state: string,
  district?: string,
  party: string,
  phone: string,
};

export type LegislatorRecord = {
  id: LegislatorIdRecord,
  identifier?: string,
  name: string | LegislatorNameRecord,
  bio: LegislatorBioRecord,
  channels: Array<LegislatorChannelRecord>,
  photoUrl?: string,
  terms: Array<LegislatorTermRecord>,
};

/**
 * Defines the `PropTypes` validation structure for a `Legislator` record.
 */
export const PropType = PropTypes.shape({
  id: PropTypes.shape({
    bioguide: PropTypes.string,
    fec: PropTypes.arrayOf(PropTypes.string),
  }),
  identifier: PropTypes.string,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      first: PropTypes.string,
      last: PropTypes.string,
      official_full: PropTypes.string.isRequired,
    }),
  ]),
  bio: PropTypes.shape({
    birthday: PropTypes.string,
    gender: PropTypes.string,
  }),
  channels: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      type: PropTypes.string,
    })
  ),
  photoUrl: PropTypes.string,
  terms: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string.isRequired,
      start: PropTypes.string.isRequired,
      end: PropTypes.string.isRequired,
      state: PropTypes.string,
      district: PropTypes.number,
      party: PropTypes.string,
    })
  ),
});

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator {
  /**
   * Get the legislator's identifier from the given `record`.
   * @param {LegislatorRecord} record that matches the `PropType` defined for
   *    `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: LegislatorRecord): LegislatorIdentifier {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator#getIdentifier'
    );
    let identifier = undefined;
    if (record.identifier !== undefined) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else {
      identifier = stripDiacritics(record.name.official_full);
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {LegislatorRecord|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<ContributionRecord>,
    legislator: LegislatorRecord | Legislator
  ): number {
    PropTypes.checkPropTypes(
      { contributions: PropTypes.arrayOf(ContributionType) },
      { contributions },
      'contributions',
      'Legislator#getContributionAmount'
    );
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  record: LegislatorRecord;

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {LegislatorRecord} record containing data about the legislator.
   */
  constructor(record: LegislatorRecord) {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator'
    );
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): LegislatorBioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): LegislatorChamber {
    if (this.isSenator) {
      return 'senate';
    } else if (this.isRepresentative) {
      return 'house';
    } else {
      return 'unknown';
    }
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): LegislatorTermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): LegislatorIdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): LegislatorIdentifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): LegislatorNameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): string {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<LegislatorTermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): ?string {
    const channels = this.record.channels || [];
    const channel = channels.find(channel => channel.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
