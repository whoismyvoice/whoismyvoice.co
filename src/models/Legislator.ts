import { Contribution } from './Contribution';
import { Record as Official } from './Official';

export type TermType = 'sen' | 'rep';
export enum Party {
  Democrat = 'Democrat',
  Republican = 'Republican',
  Independent = 'Independent',
}
export type BioguideId = string;
export type FecId = string;
export enum Chamber {
  Senate,
  House,
  Unknown,
}
export type Identifier = string;

export interface BioRecord {
  birthday: string;
  gender: string;
}

export interface ChannelRecord {
  id: string;
  type: string;
}

export interface IdRecord {
  bioguide: BioguideId;
  fec: Array<FecId>;
}

export interface NameRecord {
  first?: string;
  last?: string;
  official_full: string;
}

export interface TermRecord {
  type: TermType;
  start: string;
  end: string;
  state: string;
  district?: string;
  party: Party;
  phone: string;
}

export interface Record {
  id: IdRecord;
  identifier?: string;
  name: string | NameRecord;
  bio: BioRecord;
  channels: Array<ChannelRecord>;
  photoUrl?: string;
  terms: Array<TermRecord>;
}

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str: string) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator implements Record {
  /**
   * Holds the Record used to retrieve data for getters.
   */
  private record: Record;

  /**
   * Get the legislator's identifier from the given `record`.
   * @param {Record | Official | Legislator} record that matches the `PropType`
   *    defined for `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: Record | Official | Legislator): Identifier {
    let identifier = undefined;
    if (record instanceof Legislator) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else if (typeof record.name !== 'undefined') {
      identifier = stripDiacritics(record.name.official_full);
    } else {
      identifier = record.identifier || 'unknown';
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {Record|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<Contribution>,
    legislator: Record | Legislator
  ): number {
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {Record} record containing data about the legislator.
   */
  constructor(record: Record) {
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): BioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): Chamber {
    if (this.isSenator) {
      return Chamber.Senate;
    } else if (this.isRepresentative) {
      return Chamber.House;
    } else {
      return Chamber.Unknown;
    }
  }

  /**
   * Retrieve the channels from by which the legislator can be reached.
   * @returns an array of channels.
   */
  get channels(): Array<ChannelRecord> {
    return this.record.channels;
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): TermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): IdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): Identifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): NameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): Party {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<TermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): string | undefined {
    const channels = this.record.channels || [];
    const channel = channels.find(c => c.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
