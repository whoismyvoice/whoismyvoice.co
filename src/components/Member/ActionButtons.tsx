import * as React from 'react';

// Components
import ContactButtonSmall from './../Buttons/ContactButtonSmall';
import { Legislator } from '../../models/Legislator';

// Styles
import './../../styles/ActionButtons.scss';

interface Props {
  legislator: Legislator;
}

class SupportActions extends React.Component<Props> {
  renderTwitter() {
    const { legislator } = this.props;
    const { genderPronoun, twitter } = legislator;
    const twitterLink = `https://twitter.com/intent/tweet?hashtags=WhoIsMyVoice&text=@${twitter}`;
    return twitter === undefined ? null : (
      <ContactButtonSmall
        link={twitterLink}
        icon="twitter"
        text={`Tweet at ${genderPronoun}`}
      />
    );
  }

  render() {
    const { legislator } = this.props;
    // Define each value used for every member
    const { genderPronoun, phone } = legislator;
    return (
      <div className="actionButtons">
        <ContactButtonSmall
          link={`tel:${phone}`}
          icon="phone"
          text={`Call ${genderPronoun}`}
        />
        {this.renderTwitter()}
      </div>
    );
  }
}

export default SupportActions;
