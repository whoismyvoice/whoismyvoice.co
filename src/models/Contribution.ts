import { BioguideId } from './Legislator';

/**
 * Defines the type alias for a `Contribution` record.
 */
export interface Contribution {
  readonly amount: number;
  readonly legislatorId: BioguideId;
  readonly organization: string;
}

/**
 * Defines a record representing a contribution from a specified sector.
 */
export interface SectorContribution {
  readonly amount: number;
  readonly sector: string;
  readonly sectorCode: string;
}

/**
 * Associates a legislatorId with a list of `SectorContribution` records.
 */
export interface SectorContributions {
  readonly legislatorId: string;
  readonly contributions: Readonly<SectorContribution[]>;
}
