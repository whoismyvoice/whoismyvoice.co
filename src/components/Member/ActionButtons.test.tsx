import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import ActionButtons from './ActionButtons';
import { createLegislator } from '../../models/Legislator.test';
import { Legislator } from '../../models/Legislator';

it('renders without crashing', () => {
  const legislator = new Legislator(createLegislator('John Smith'));
  const { container } = render(<ActionButtons legislator={legislator} />);
  expect(container).toBeInTheDocument();
  expect(container.querySelector('.actionButtons')).toBeInTheDocument();
  expect(
    container.querySelector('.button__small__icon.phone')
  ).toBeInTheDocument();
});

it('has no twitter contact if no channels', () => {
  const legislator = new Legislator({
    ...createLegislator('John Smith'),
    channels: [],
  });
  const { container } = render(<ActionButtons legislator={legislator} />);
  expect(container.querySelector('.button__small__icon.twitter')).toBeNull();
});

it('has twitter contact if twitter channel', () => {
  const legislator = new Legislator({
    ...createLegislator('John Smith'),
    channels: [{ id: 'johnsmith', type: 'Twitter' }],
  });
  const { container } = render(<ActionButtons legislator={legislator} />);
  expect(
    container.querySelector('.button__small__icon.twitter')
  ).toBeInTheDocument();
});
