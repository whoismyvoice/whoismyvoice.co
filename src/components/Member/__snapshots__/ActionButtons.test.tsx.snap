// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="actionButtons"
>
  <a
    href="tel:123-123-1234"
    rel="noopener noreferrer"
    target="_blank"
  >
    <button
      className="contact__button_small undefined"
    >
      <div
        className="button__small__icon phone"
      />
      <div
        className="button__small__text"
      >
        Call Them
      </div>
    </button>
  </a>
</div>
`;
